integration-libs/epd-visualization/components/visual-picking/visual-picking-tab/product-list/visual-picking-product-list.service.ts

```diff
@@ -0,0 +1,171 @@
+   * @returns An Observable that produces the spare part product references for the current product.
+   */
+  public getProductReferences$(): Observable<ProductReference[]> {
+    return this.currentProduct$.pipe(
+      tap((product: Product) =>
+        this.productReferenceService.loadProductReferences(
+          product.code as string,
+          this.productReferenceType
+        )
+      ),
+      switchMap((product) =>
+        this.productReferenceService.getProductReferences(
+          product.code as string,
+          this.productReferenceType
+        )
+      ),
+      filter(
+        (productReferences: ProductReference[]) =>
+          productReferences !== undefined
+      ),
+      distinctUntilChanged((x, y) => JSON.stringify(x) === JSON.stringify(y)),
+      shareReplay()
+    );
+  }
+
+  /**
+   * Returns an Observable that produces a filtered array of spare part product references for the current product.
+   * Filtering is performed by the VisualPickingProductFilterService.
+   * @returns An Observable that produces a filtered array of spare part product references for the current product.
+   */
+  public getFilteredProductReferences$(): Observable<ProductReference[]> {
+    return this.visualPickingProductFilterService
+      .getFilteredProducts$(this.getProductReferences$())
+      .pipe(shareReplay());
+  }
+
+  public activeSlideStartIndex = 0;
+  public itemsPerSlide = this.DEFAULT_ITEMS_PER_SLIDE;
+
+  public set selectedProductCodes(selectedProductCodes: string[]) {
+    this._selectedProductCodes = selectedProductCodes;
+    this.selectedProductCodesChange.next(selectedProductCodes);
+  }
+  public get selectedProductCodes(): string[] {
+    return this._selectedProductCodes;
+  }
+  private _selectedProductCodes: string[];
+  public selectedProductCodesChange = new EventEmitter<string[]>();
+
+  /**
+   * Used to create the list item model data for the visual picking product list.
+   * Returns an observable containing an array of VisualPickingProductListItem objects created by combining the latest values from
+   * an Observable producing an array of product references and
+   * an Observable producing an array of selected product codes.
+   * The VisualPickingProductListItem model object combines a ProductReference for a spare part and the selected/highlighted state of the list item.
+   * @param productReferences$ An Observable producing the array of ProductReference values to map.
+   * @param selectedProductCodes$ An Observable producing the array of selected product codes.
+   * @returns An Observable producing an array of VisualPickingProductListItem values.
+   */
+  public getVisualPickingProductListItems$(
+    productReferences$: Observable<ProductReference[]>,
+    selectedProductCodes$: Observable<string[]>
+  ): Observable<VisualPickingProductListItem[]> {
+    return combineLatest([productReferences$, selectedProductCodes$]).pipe(
+      filter((valuePair) => !!valuePair[0] && !!valuePair[1]),
```

Comment:
hint: for readability you might want to destructure the argument of the function into array elements, e.g. `filter(([productReferences, selectedProductCodes]) => !!productReferences && !!selectedProductCodes)`. But that's not required. Just FYI

Reply:
That's nicer. I'll do that.